src/librustc_mir/transform/nrvo.rs

```diff
@@ -0,0 +1,147 @@
+use rustc_index::bit_set::HybridBitSet;
+use rustc_middle::mir::visit::{MutVisitor, PlaceContext};
+use rustc_middle::mir::{self, BasicBlock, Local, Location};
+use rustc_middle::ty::TyCtxt;
+
+use crate::transform::{MirPass, MirSource};
+
+pub struct RenameReturnPlace;
+
+/// Renames the local corresponding to the
+impl<'tcx> MirPass<'tcx> for RenameReturnPlace {
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, src: MirSource<'tcx>, body: &mut mir::Body<'tcx>) {
+        /*
+        if tcx.sess.opts.debugging_opts.mir_opt_level == 2 {
+            return;
+        }
+        */
+
+        debug!("Determining if `{:?}` is eligible for NRVO...", src.def_id());
+        let returned_local = match single_local_assigned_to_return_place(body) {
+            Some(l) => l,
+            None => {
+                debug!("MIR was ineligible for NRVO");
+                return;
+            }
+        };
+
+        debug!("MIR was eligible for NRVO, making {:?} the return place", returned_local);
+
+        RenameToReturnPlace { tcx, orig: returned_local }.visit_body(body);
+
+        for block_data in body.basic_blocks_mut() {
+            block_data.statements.retain(|stmt| stmt.kind != mir::StatementKind::Nop);
+        }
+    }
+}
+
+/// Looks at all basic blocks that are terminated with a `Return` to see what local is copied to
+/// the return place in that basic block. If all `Return`-terminated blocks assign the same local
+/// to the return place, return that local.
+fn single_local_assigned_to_return_place(body: &mut mir::Body<'_>) -> Option<Local> {
```

Comment:
If there are any uses of `_0` without it being a projectionless place on the lhs of an assignment, we should also return `None`. This may not be a problem right now, but it may very well occur after a combination of applying this optimization, then inlining the function the optimization was applied to, and then applying the optimization on the outer function.

Reply:
This is resolved via the `IsReturnPlaceUsed` visitor. It also will fire for assignments of projections of `_0`, although that's probably not obvious on first glance. I kinda want to change `PlaceContext` to remain unchanged instead of falling back to `Projection` when we see something like `_1.field`. I'd want to introduce a `PlaceContext::Deref` though.